bigquery_etl/stripe/allowed_fields.py

```diff
@@ -0,0 +1,188 @@
+"""Limit imported Stripe data to a set of allowed fields."""
+
+import re
+from hashlib import sha256
+from pathlib import Path
+from typing import Any, Generator, Iterable, List, Type
+
+import click
+import stripe
+import ujson
+import yaml
+from google.cloud import bigquery
+from stripe.api_resources.abstract import ListableAPIResource
+
+
+def _snake_case(resource: Type[ListableAPIResource]) -> str:
+    return re.sub(r"(?<!^)(?=[A-Z])", "_", resource.__name__).lower()
+
+
+# event data types with separate events and a defined schema
+ALLOWED_FIELDS = yaml.load((Path(__file__).parent / "allowed_fields.yaml").read_text())
+
+
+def _valid_float(value: str):
+    try:
+        float(value)
+    except ValueError:
+        return False
+    else:
+        return True
+
+
+class FilteredSchema:
+    """Apply ALLOWED_FIELDS to resources and their schema."""
+
+    type: str
+    allowed: dict
+    root: bigquery.SchemaField
+    filtered: List[bigquery.SchemaField]
+
+    def __init__(self, resource: Type[ListableAPIResource]):
+        """Get filtered schema and allowed fields for a Stripe resource type."""
+        self.type = _snake_case(resource)
+        self.allowed = ALLOWED_FIELDS[self.type]
+        path = Path(__file__).parent / f"{self.type}.schema.json"
+        self.root = bigquery.SchemaField.from_api_repr(
+            {"name": "root", "type": "RECORD", "fields": ujson.loads(path.read_text())}
+        )
+        self.filtered = list(self._filter_schema(self.root.fields, self.allowed))
+
+    def _filter_schema(
+        self, fields: List[bigquery.SchemaField], allowed: dict
+    ) -> Iterable[bigquery.SchemaField]:
+        for field in fields:
+            if field.name in allowed:
+                if field.field_type != "RECORD":
+                    yield field
+                else:
+                    yield bigquery.SchemaField.from_api_repr(
+                        {
+                            **field.to_api_repr(),
+                            fields: list(
+                                self._filter_schema(field.fields, allowed[field.name])
+                            ),
+                        }
+                    )
+
+    def format_row(self, row: dict) -> bytes:
+        """Format stripe object for BigQuery, and validate against original schema."""
+        return ujson.dumps(
+            self._format_helper(
+                self.type, obj=row, allowed=self.allowed, field=self.root
+            )
+        ).encode("UTF-8")
+
+    def _format_helper(
+        self,
+        *path,
+        obj: Any,
+        allowed: dict,
+        field: bigquery.SchemaField,
+        is_list_item: bool = False,
+    ) -> Any:
+        # override obj for metadata and paged lists
+        if (
+            field.name in ("metadata", "custom_fields")
+            and isinstance(obj, dict)
+            and not is_list_item
+        ):
+            if "userid" in obj:
+                # hash fxa uid before it reaches BigQuery
```

Comment:
Why?

Reply:
@jklukas told me we don't want unhashed fxa uid anywhere in `moz-fx-data-shared-prod`

Reply:
If the relevant tables here are Workgroup Confidential, then we could consider leaving this unhashed. But I'd indeed like to keep raw fxa userid out of shared-prod unless we have a clear business case.

Reply:
we don't, this is working fine